### PR TITLE
CBG-880: SGCollect Python3 compatibility

### PR DESCRIPTION
**Pull request by JRascagneres**
Seems to work fine. Tested following scenarios:

Sync Gateway Running
- Normal run
- Redacted run
- Redacted with salt run
- Upload

Sync Gateway Not Running
- Normal run
- Redacted run
- Redacted with salt run
- Upload

### CODE REVIEW

File: tools/tasks.py [hunk 1]
```diff
@@ -14,10 +14,10 @@
 import optparse
 import atexit
 import signal
-import urllib
+import urllib.request, urllib.parse, urllib.error
 import shutil
-import urlparse
-import urllib2
+import urllib.parse
+import urllib.request, urllib.error, urllib.parse
```
Comment (bbrks) [inline tools/tasks.py]: duplicate imports?

File: tools/tasks.py [hunk 2]
```diff
@@ -217,14 +216,17 @@ def __init__(self, description, callable, timeout=None, **kwargs):
...
-            fp.write(result)
+            try:
+                fp.write(result.encode())
+            except (UnicodeEncodeError, AttributeError):
+                fp.write(result)
             return 0
         except Exception as e:
             if self.log_exception:
-                print("Exception executing python task: {0}".format(e))
+                print(("Exception executing python task: {0}".format(e)))
```
Comment (bbrks) [inline tools/tasks.py]: Almost all of the prints containing only a string don't need double brackets. See example on line 255

**Review by bbrks — CHANGES_REQUESTED** [review]
2 comments but would also be worth spot checking cbcollect_info to see what changes they made to bits of code we share (e.g. log redaction)

File: tools/tasks.py [hunk 3]
```diff
@@ -217,10 +216,13 @@ def __init__(self, description, callable, timeout=None, **kwargs):
 
     def execute(self, fp):
         """Run the task"""
-        print("log_file: {0}. ".format(self.log_file))
+        print(("log_file: {0}. ".format(self.log_file)))
```
Comment (adamcfraser) [inline tools/tasks.py]: Are the double parentheses needed here?

File: tools/tasks.py [hunk 4]
```diff
@@ -217,10 +216,13 @@ def __init__(self, description, callable, timeout=None, **kwargs):
...
         try:
             result = self.callable()
-            fp.write(result)
+            try:
+                fp.write(result.encode())
+            except (UnicodeEncodeError, AttributeError):
+                fp.write(result)
```
Comment (adamcfraser) [inline tools/tasks.py:225]: If result.encode() has an error, what is fp.write(result) going to do?

File: tools/tasks.py [hunk 5]
```diff
@@ -479,15 +478,15 @@ def add_gzip_file_task(sourcefile_path, salt, content_postprocessors=[]):
...
             for content_postprocessor in content_postprocessors:
                 contents = content_postprocessor(contents)
             redactor = LogRedactor(salt, tempfile.mkdtemp())
             contents = redactor.redact_string(contents)
 
-        out = StringIO.StringIO()
+        out = io.BytesIO()
         with gzip.GzipFile(fileobj=out, mode="w") as f:
-            f.write(contents)
+            f.write(contents.encode())
```
Comment (adamcfraser) [inline tools/tasks.py:489]: Elsewhere you're doing try/catch for errors on encode() - should that be happening here?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Generally looks good - a few minor questions/clarifications.

File: tools/tasks.py (hunk 3 above)
Comment (JRascagneres) [reply tools/tasks.py]: Ah nope, thanks. Thought I'd caught those.

File: tools/tasks.py (hunk 4 above)
Comment (JRascagneres) [reply tools/tasks.py:225]: Its going to send the result as is. This is due to the fact that sometimes result is a string and sometimes bytes. This handles this case where it'll try to convert to bytes, however, if its already in byte format then it'll fall into the except case.

Comment (JRascagneres) [reply tools/tasks.py:225]: It needs to be in bytes to save it

File: tools/tasks.py (hunk 5 above)
Comment (JRascagneres) [reply tools/tasks.py:489]: No. We know the contents came from file and know we've converted them to string and need encoding back.

File: tools/tasks.py (hunk 3 above)
Comment (JRascagneres) [reply tools/tasks.py]: Fixed :)

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**